// clumber/lib/lumber/transports/file.js
// Generated by CoffeeScript 1.3.3

/*
@license
file.js: File transport

Liberal inspiration taken from https://github.com/flatiron/winston

(c) 2012 Panther Development
MIT LICENSE
*/


(function() {
  var File, dateFormat, events, fs, lumber, path, util,
    __bind = function(fn, me){ return function(){ return fn.apply(me, arguments); }; },
    __hasProp = {}.hasOwnProperty,
    __extends = function(child, parent) { for (var key in parent) { if (__hasProp.call(parent, key)) child[key] = parent[key]; } function ctor() { this.constructor = child; } ctor.prototype = parent.prototype; child.prototype = new ctor(); child.__super__ = parent.prototype; return child; };

  util = require("util");

  fs = require("fs");

  path = require("path");

  events = require("events");

  dateFormat = require("dateformat");

  lumber = require("../../lumber");

  /*
  File Transport
  @constructor
  @implements {Transport}
  */


  File = (function(_super) {

    __extends(File, _super);

    File._todaysDate = void 0;

    function File(options) {
      var e;
      if (options == null) {
        options = {};
      }
      this._flush = __bind(this._flush, this);

      File.__super__.constructor.call(this);
      this.encoder = lumber.util.checkOption(options.encoder, "json");
      this.level = lumber.util.checkOption(options.level, "info");
      this.filename = lumber.util.checkOption(options.filename, path.resolve("app.log"));
      this.filemode = lumber.util.checkOption(options.filemode, "0666");
      this.rotate = lumber.util.checkOption(options.rotate, false);
      this._rotating = false;
      this._buffer = [];
      this.name = "file";
      if (typeof this.encoder === "string") {
        e = lumber.util.titleCase(this.encoder);
        if (lumber.encoders[e]) {
          this.encoder = new lumber.encoders[e]();
        } else {
          throw new Error("Unknown encoder passed: " + this.encoder);
        }
      }
      this.encoding = this.encoder.encoding;
      this._todaysDate = new Date();
    }

    File.prototype.log = function(args, cb) {
      var msg,
        _this = this;
      msg = this.encoder.encode(args.level, args.msg, args.meta);
      return this._open(function(buff) {
        if (buff || _this._rotating) {
          return _this._buffer.push([msg, args, cb]);
        } else {
          return _this._write(msg + "\n", function(err) {
            if (cb) {
              return cb(err, msg, args.level, _this.name, _this.filename);
            }
          });
        }
      });
    };

    File.prototype._write = function(data, cb) {
      var _this = this;
      if (this._stream.write(data, this.encoding)) {
        return this._rotateLogsIfNecessary(cb);
      } else {
        return this._drain(function() {
          return _this._rotateLogsIfNecessary(cb);
        });
      }
    };

    File.prototype._rotateLogsIfNecessary = function(cb) {
      var _this = this;
      if (this._needsToRotateLogs()) {
        return this._rotateLogs(function(err) {
          _this._todaysDate = new Date();
          if (cb) {
            return cb(err);
          }
        });
      } else {
        if (cb) {
          return cb(null);
        }
      }
    };

    File.prototype._open = function(cb) {
      var _this = this;
      if (this._opening) {
        if (cb) {
          return cb(true);
        }
      } else if (this._stream) {
        if (cb) {
          return cb(false);
        }
      } else {
        if (cb) {
          cb(true);
        }
        this._stream = fs.createWriteStream(this.filename, {
          flags: "a",
          encoding: this.encoding,
          mode: this.fileMode
        });
        this._stream.setMaxListeners(Infinity);
        this.once("flush", function() {
          _this._opening = false;
          return _this.emit("open", _this.filename);
        });
        return this._flush();
      }
    };

    File.prototype._close = function(cb) {
      if (this._stream) {
        this._stream.on('close', function() {
          this.emit("closed");
          if (cb) {
            return cb(null);
          }
        });
        this._stream.end();
        this._stream.destroySoon();
        return this._stream = null;
      } else {
        this._stream = null;
        if (cb) {
          return cb(null);
        }
      }
    };

    File.prototype._drain = function(cb) {
      var _this = this;
      return this._stream.once("drain", function() {
        _this.emit("drain");
        if (cb) {
          return cb();
        }
      });
    };

    File.prototype._flush = function(cb) {
      var _this = this;
      if (this._buffer.length === 0) {
        this.emit("flush");
        if (cb) {
          cb(null);
        }
        return;
      }
      this._buffer.forEach(function(log) {
        var args, msg;
        msg = log[0], args = log[1], cb = log[2];
        return process.nextTick(function() {
          return _this._write(msg + "\n", function(err) {
            if (cb) {
              return cb(err, msg, args.level, _this.name, _this.filename);
            }
          });
        });
      });
      this._buffer.length = 0;
      return this._drain(function() {
        _this.emit("flush");
        if (cb) {
          return cb(null);
        }
      });
    };

    File.prototype._needsToRotateLogs = function() {
      var d, shouldIRotate;
      d = new Date();
      shouldIRotate = !(d.getDate() === this._todaysDate.getDate() && d.getMonth() === this._todaysDate.getMonth() && d.getYear() === this._todaysDate.getYear());
      return !this._rotating && this.rotate && shouldIRotate;
    };

    File.prototype._rotateLogs = function(cb) {
      var _this = this;
      this._rotating = true;
      return this._close(function() {
        var from, to;
        from = _this.filename;
        to = _this.filename + "." + dateFormat(_this._todaysDate, 'mm-dd-yyyy');
        return fs.rename(from, to, function(err) {
          _this._rotating = false;
          if (cb && err) {
            return cb(err);
          }
          _this.emit('rotate');
          _this._flush();
          return cb();
        });
      });
    };

    return File;

  })(events.EventEmitter);

  module.exports.File = File;

}).call(this);
